databases/core.py

```diff
@@ -156,6 +160,21 @@ def __init__(self, backend: DatabaseBackend) -> None:
             if self._connection_counter == 0:
                 await self._connection.release()
 
+    def __call__(self, func: typing.Callable) -> typing.Callable:
+        """
+        Called if using `@database.connection()` as a decorator.
+        """
+        @functools.wraps(func)
+        async def wrapper(*args: typing.Any, **kwargs: typing.Any) -> typing.Any:
+            connected_by_decorator = False
+            if not self._backend._pool:
+                await self._backend.connect()
+                connected_by_decorator = True
+
+            async with self:
+                return await func(*args, **kwargs)
+        return wrapper
```

Comment:
Having this logic within the decorator isn't very nice, plus there's some ambiguity where the db should be disconnected back, which makes it quite expensive operation though.